org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -213,24 +217,18 @@ SynchronizationResult run(CancellationTokenSource tokenSource, IProgressMonitor
 
         @Override
         public void runInToolingApi(CancellationTokenSource tokenSource, IProgressMonitor monitor) throws Exception {
-
             try {
                 SubMonitor progress = SubMonitor.convert(monitor, 5);
                 progress.setTaskName((String.format("Synchronizing Gradle build at %s with workspace", this.gradleBuild.getBuildConfig().getRootProjectDirectory())));
                 new ImportRootProjectOperation(this.gradleBuild.getBuildConfig(), this.newProjectHandler).run(progress.newChild(1));
-                Collection<EclipseProject> rootProjects = ModelProviderUtil.fetchAllEclipseProjectsWithSyncTask(this.gradleBuild, tokenSource, FetchStrategy.FORCE_RELOAD, progress.newChild(1));
-                Set<EclipseProject> allProjects = collectAll(rootProjects);
+                Set<EclipseProject> allProjects = collectAll(this.gradleBuild.modelProvider.fetchEclipseProjectAndRunSyncTasks(tokenSource, progress.newChild(1)));
                 new ValidateProjectLocationOperation(allProjects).run(progress.newChild(1));
                 new RunOnImportTasksOperation(allProjects, this.gradleBuild.getBuildConfig()).run(progress.newChild(1), tokenSource);
                 this.failures = new SynchronizeGradleBuildOperation(allProjects, this.gradleBuild, this.newProjectHandler,
                         ProjectConfigurators.create(this.gradleBuild, CorePlugin.extensionManager().loadConfigurators())).run(progress.newChild(1));
-
-                // The root projects are no longer loaded via GradleBuild.getModelProvider(). To avoid unnecessary
-                // model loads (e.g. when somebody opens the test sources from the Execution view) we add the result
-                // to the cache here.
-                ((DefaultModelProvider)this.gradleBuild.getModelProvider()).getCache().put(EclipseProject.class, rootProjects);
             } finally {
                 this.gradleBuild.projectConnectionCache.invalidateAll();
+                IdeFriendlyClassLoading.cleanup();
```

Comment:
I feel like this is the wrong place for this - Shouldn't this be done when our plugin is unloaded instead?

Reply:
Let's do that.